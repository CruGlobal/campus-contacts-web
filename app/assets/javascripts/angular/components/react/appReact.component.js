import template from './appReact.html';

angular.module('missionhubApp').component('appReact', {
    controller: AppReactController,
    template: template,
});

function AppReactController($rootScope) {}
