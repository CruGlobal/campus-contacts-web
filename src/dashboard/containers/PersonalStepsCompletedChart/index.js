import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useQuery } from 'react-apollo-hooks';
import gql from 'graphql-tag';
import styled from '@emotion/styled';
import { withTheme } from 'emotion-theming';
import _ from 'lodash';
import moment from 'moment';

import BarChart from '../../components/BarChart';
import SwitchButton from '../../components/SwitchButton';

const GET_STEPS_COMPLETED_REPORT = gql`
    query communityReport {
        communityReport {
            dayReport {
                date
                stepsCompletedCount
                communityStagesReport {
                    pathwayStage
                    stepsCompletedCount
                }
            }
        }
    }
`;

const Wrapper = styled.div`
    height: 400px;
    position: relative;
`;

const Footer = styled.div`
    height: 32px;
    display: flex;
    justify-content: space-between;
`;

const TooltipRow = styled.div`
    display: flex;
    justify-content: space-between;

    &.total {
        color: ${({ theme }) => theme.colors.highlight};
    }
`;

const Line = styled.div`
    border-top: 2px solid ${({ theme }) => theme.colors.highlight};
    width: 26px;
    height: 2px;

    &.dashed {
        border-top-style: dashed;
    }
`;

const Legend = styled.div`
    display: flex;
    align-items: center;
`;

const LegendLabel = styled.div`
    font-size: 12px;
    line-height: 16px;
    color: ${({ theme }) => theme.colors.secondary};
    margin-left: 5px;
    margin-right: 10px;
`;

const Label = styled.div`
    font-weight: 600;
    font-size: 10px;
    line-height: 20px;
    text-transform: uppercase;
    letter-spacing: 1px;
    margin-right: 20px;
`;

const Value = styled.div`
    font-weight: 600;
    font-size: 10px;
    line-height: 20px;
    text-transform: uppercase;
    letter-spacing: 1px;
    align-self: flex-end;
`;

const PreviousButton = styled.div`
    width: 6px;
    height: 12px;
    border-top: 6px solid transparent;
    border-bottom: 6px solid transparent;
    border-right: 6px solid #b4b6ba;
    position: absolute;
    bottom: 57px;
    left: 19px;
    cursor: pointer;

    &:hover {
        border-right: 6px solid #c3c3c3;
    }
`;

const NextButton = styled.div`
    width: 6px;
    height: 12px;
    border-top: 6px solid transparent;
    border-bottom: 6px solid transparent;
    border-left: 6px solid #b4b6ba;
    position: absolute;
    bottom: 57px;
    right: -10px;
    cursor: pointer;

    &:hover {
        border-left: 6px solid #c3c3c3;
    }
`;

const PersonalStepsCompletedChart = () => {
    const [filter, setFilter] = useState({
        type: 'month',
        index: 0,
        datesRange: [moment().subtract(1, 'month'), moment()],
    });

    const { data, loading } = useQuery(GET_STEPS_COMPLETED_REPORT);
    const { t } = useTranslation('insights');

    if (loading) {
        return <Wrapper>{t('loading')}</Wrapper>;
    }

    // To be used when quering data from GraphQL Endpoint
    // console.log('From:', filter.datesRange[0].format('YYYY-MM-DD'));
    // console.log('To:', filter.datesRange[1].format('YYYY-MM-DD'));

    const updateDateRange = (index, type) => {
        setFilter({
            type,
            index,
            datesRange: [
                moment().subtract(index + 1, type),
                moment().subtract(index, type),
            ],
        });
    };

    const onToggleFilterClick = filterType => {
        updateDateRange(0, filterType);
    };

    const onNextRangeClick = () => {
        const newIndex = filter.index - 1 >= 0 ? filter.index - 1 : 0;
        updateDateRange(newIndex, filter.type);
    };

    const onPreviousRangeClick = () => {
        const newIndex = filter.index + 1;
        updateDateRange(newIndex, filter.type);
    };

    const {
        communityReport: { dayReport: report },
    } = data;

    const tooltip = options => {
        const { data } = options;

        return [
            data.stages.map(row => (
                <TooltipRow key={row.name}>
                    <Label>{row.name}</Label>
                    <Value>{row.count}</Value>
                </TooltipRow>
            )),
            <TooltipRow className={'total'} key={'total'}>
                <Label>Total</Label>
                <Value>{data.stepsTotal}</Value>
            </TooltipRow>,
        ];
    };

    const graphData = report.map(row => ({
        ['stepsTotal']: row.stepsCompletedCount,
        ['stages']: row.communityStagesReport.map(stage => ({
            name: stage.pathwayStage,
            count: stage.stepsCompletedCount,
        })),
        ['date']: row.date.substring(0, 2),
    }));

    const average = Math.floor(
        _.sumBy(graphData, 'stepsTotal') / graphData.length,
    );

    return (
        <div>
            <Wrapper>
                <BarChart
                    data={graphData}
                    keys={['stepsTotal']}
                    indexBy={'date'}
                    average={average}
                    tooltip={tooltip}
                />
                <PreviousButton onClick={onPreviousRangeClick} />
                <NextButton onClick={onNextRangeClick} />
            </Wrapper>
            <Footer>
                <SwitchButton
                    leftLabel={t('monthLabel')}
                    rightLabel={t('yearLabel')}
                    onLeftClick={() => onToggleFilterClick('month')}
                    onRightClick={() => onToggleFilterClick('year')}
                />
                <Legend>
                    <Line className={'dashed'} />
                    <LegendLabel>Average</LegendLabel>
                    <Line />
                    <LegendLabel>Steps of Faith</LegendLabel>
                </Legend>
            </Footer>
        </div>
    );
};

export default withTheme(PersonalStepsCompletedChart);
