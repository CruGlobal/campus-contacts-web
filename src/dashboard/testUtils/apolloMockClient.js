import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
