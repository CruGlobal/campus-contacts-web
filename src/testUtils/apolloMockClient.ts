import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
// @ts-ignore
import { buildClientSchema } from 'graphql/utilities';
// @ts-ignore
import { createMockerFromIntrospection } from 'fraql/mock';
// @ts-ignore
import gql from 'fraql';

import introspectionQuery from '../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument: any, options: any) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
