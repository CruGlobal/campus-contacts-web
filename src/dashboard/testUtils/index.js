import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
