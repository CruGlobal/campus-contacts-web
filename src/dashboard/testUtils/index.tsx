import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
