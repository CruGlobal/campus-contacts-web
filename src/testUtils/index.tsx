import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
