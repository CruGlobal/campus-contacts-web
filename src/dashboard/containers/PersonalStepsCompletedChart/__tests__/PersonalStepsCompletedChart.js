// Vendors
import React from 'react';
import { waitForElement } from '@testing-library/react';

import { renderWithContext } from '../../../testUtils';
import PersonalStepsCompletedChart from '../';

describe('<PersonalStepsCompletedChart />', () => {
    it('should render properly loading state', async () => {
        renderWithContext(<PersonalStepsCompletedChart />).snapshot();
    });

    // TODO: Add unit tests when communityReport is available in GraphQL schema
});
