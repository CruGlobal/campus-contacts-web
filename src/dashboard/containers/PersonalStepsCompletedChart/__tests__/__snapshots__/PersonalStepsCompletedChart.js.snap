// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonalStepsCompletedChart /> should render properly loading state 1`] = `
<div>
  <div
    class="css-1a3wyyp"
  >
    Loading...
  </div>
</div>
`;
